{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To dos:\n",
    "- create a dataset with age and lon, lat and download it - CHECK\n",
    "- specify a new location to save images on the server -  CHECK\n",
    "- loop over the dataset - CHECK\n",
    " - plug in lat and lon into the api code below - CHECK\n",
    " - extract all images for that location - CHECK\n",
    " - name image according to age and its unique code after the last - CHECK\n",
    "\n",
    "CNN\n",
    "- rewatch simple CNN video how to extract age from name file\n",
    "- run a simple CNN and compare results\n",
    "\n",
    "Extent dataset:\n",
    "- access Mapillary data either through API or the download you did\n",
    "- repeat above excercise\n",
    "- compare R2\n",
    "- compare coverage of images between the two - are the duplicates?"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Imports"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "import requests\n",
    "import re\n",
    "import json\n",
    "import os\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Loaded processed DataFrame from /workspace/workspace/ufo-prediction/demo/kartaview_key.csv\n",
      "           lon        lat  age_right\n",
      "0    -3.688602  40.459993     1967.0\n",
      "1    -3.687594  40.433276     1977.0\n",
      "2     2.159530  41.392549     1890.0\n",
      "3    -3.710879  40.383638     1962.0\n",
      "4    -3.618895  40.457296     2005.0\n",
      "...        ...        ...        ...\n",
      "1340 -0.570843  38.340551     2002.0\n",
      "1341 -8.404795  43.339535     2004.0\n",
      "1342  2.173094  41.409972     1977.0\n",
      "1343  2.155593  41.375314     2012.0\n",
      "1344  2.171494  41.377807     1982.0\n",
      "\n",
      "[1345 rows x 3 columns]\n"
     ]
    }
   ],
   "source": [
    "# Define the path for the processed DataFrame\n",
    "processed_df_path = \"/workspace/workspace/ufo-prediction/demo/kartaview_key.csv\"\n",
    "\n",
    "# Load the existing processed DataFrame if it exists, else create a new one\n",
    "if os.path.exists(processed_df_path):\n",
    "    kartaview_keys = pd.read_csv(processed_df_path)\n",
    "    print(\"Loaded processed DataFrame from\", processed_df_path)\n",
    "else:\n",
    "    # Assuming df_subset is initially loaded or defined here\n",
    "   print(\"Creating a new processed DataFrame\")\n",
    "   path_data_RCA = os.path.join(dataset.DATA_DIR, 'rca-ufo-merge_ALL.csv')\n",
    "   df = pd.read_csv(path_data_RCA, encoding='latin1')\n",
    "   kartaview_keys = df[['lon', 'lat','age_right']]\n",
    "   kartaview_keys.to_csv('/workspace/workspace/ufo-prediction/demo/kartaview_key.csv', index=False)\n",
    "\n",
    "print(kartaview_keys)   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory /workspace/workspace/ufo-prediction/image_data already exists.\n"
     ]
    }
   ],
   "source": [
    "# Set directory\n",
    "\n",
    "current_directory = '/workspace/workspace/ufo-prediction/image_data'\n",
    "\n",
    "# Check if the new directory exists, if not, create it\n",
    "if not os.path.exists(current_directory):\n",
    "    os.makedirs(current_directory)\n",
    "    print(f\"Directory {current_directory} created.\")\n",
    "else:\n",
    "    print(f\"Directory {current_directory} already exists.\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "No suitable images found for location: lon=-3.688602, lat=40.459993\n",
      "No suitable images found for location: lon=-3.68860, lat=40.45999\n",
      "No suitable images found for location: lon=-3.6886, lat=40.4600\n",
      "No suitable images found for location: lon=-3.689, lat=40.460\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.6886020866307896, lat=40.45999341644553\n",
      "No suitable images found for location: lon=-3.687594, lat=40.433276\n",
      "No suitable images found for location: lon=-3.68759, lat=40.43328\n",
      "No suitable images found for location: lon=-3.6876, lat=40.4333\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1977.0_2799522_668a1_5f18beaca98d7.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1977.0_3484289_3bb7ecaf4e38c839ab9fbdf2f898591d.jpg\n",
      "No suitable images found for location: lon=2.159530, lat=41.392549\n",
      "No suitable images found for location: lon=2.15953, lat=41.39255\n",
      "No suitable images found for location: lon=2.1595, lat=41.3925\n",
      "No suitable images found for location: lon=2.160, lat=41.393\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=2.159529927925693, lat=41.39254926659798\n",
      "No suitable images found for location: lon=-3.710879, lat=40.383638\n",
      "No suitable images found for location: lon=-3.71088, lat=40.38364\n",
      "No suitable images found for location: lon=-3.7109, lat=40.3836\n",
      "No suitable images found for location: lon=-3.711, lat=40.384\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.710878872855944, lat=40.38363790788449\n",
      "No suitable images found for location: lon=-3.618895, lat=40.457296\n",
      "No suitable images found for location: lon=-3.61890, lat=40.45730\n",
      "No suitable images found for location: lon=-3.6189, lat=40.4573\n",
      "No suitable images found for location: lon=-3.619, lat=40.457\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.618895090865216, lat=40.4572955068167\n",
      "No suitable images found for location: lon=-3.687021, lat=40.433321\n",
      "No suitable images found for location: lon=-3.68702, lat=40.43332\n",
      "No suitable images found for location: lon=-3.6870, lat=40.4333\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1975.0_2799522_9acdf_5f18bf6fd7220.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1975.0_3484289_b24faa474e07d1f709c53c417d4f06e1.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1917.0_1955509_22_76d9b_7.jpg\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1917.0_3714237_88d34738c613f0f1110d618ec285d86a.jpg. Skipping download.\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1917.0_6185097_77989f38d1f499f687dbc1749eddbc22.jpg. Skipping download.\n",
      "No suitable images found for location: lon=-3.682960, lat=40.430620\n",
      "No suitable images found for location: lon=-3.68296, lat=40.43062\n",
      "No suitable images found for location: lon=-3.6830, lat=40.4306\n",
      "No suitable images found for location: lon=-3.683, lat=40.431\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.682959519781954, lat=40.43062014180322\n",
      "No suitable images found for location: lon=-6.997229, lat=38.883504\n",
      "No suitable images found for location: lon=-6.99723, lat=38.88350\n",
      "No suitable images found for location: lon=-6.9972, lat=38.8835\n",
      "No suitable images found for location: lon=-6.997, lat=38.884\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-6.997228686265334, lat=38.883503824248166\n",
      "No suitable images found for location: lon=-3.620598, lat=40.437601\n",
      "No suitable images found for location: lon=-3.62060, lat=40.43760\n",
      "No suitable images found for location: lon=-3.6206, lat=40.4376\n",
      "No suitable images found for location: lon=-3.621, lat=40.438\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.6205977979501496, lat=40.43760091243603\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1970.0_3714237_6c53d21ddbc2562fdb3112a3de30f86d.jpg. Skipping download.\n",
      "No suitable images found for location: lon=-3.698112, lat=40.238291\n",
      "No suitable images found for location: lon=-3.69811, lat=40.23829\n",
      "No suitable images found for location: lon=-3.6981, lat=40.2383\n",
      "No suitable images found for location: lon=-3.698, lat=40.238\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.698111674526544, lat=40.23829101690725\n",
      "No suitable images found for location: lon=-3.626364, lat=40.547549\n",
      "No suitable images found for location: lon=-3.62636, lat=40.54755\n",
      "No suitable images found for location: lon=-3.6264, lat=40.5475\n",
      "No suitable images found for location: lon=-3.626, lat=40.548\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.626364258446926, lat=40.54754923957008\n",
      "No suitable images found for location: lon=-3.632043, lat=40.434859\n",
      "No suitable images found for location: lon=-3.63204, lat=40.43486\n",
      "No suitable images found for location: lon=-3.6320, lat=40.4349\n",
      "No suitable images found for location: lon=-3.632, lat=40.435\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.6320426821469423, lat=40.43485909014028\n",
      "No suitable images found for location: lon=-3.693448, lat=40.434132\n",
      "No suitable images found for location: lon=-3.69345, lat=40.43413\n",
      "No suitable images found for location: lon=-3.6934, lat=40.4341\n",
      "No suitable images found for location: lon=-3.693, lat=40.434\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.693447726031988, lat=40.43413210403984\n",
      "No suitable images found for location: lon=-3.631763, lat=40.438113\n",
      "No suitable images found for location: lon=-3.63176, lat=40.43811\n",
      "No suitable images found for location: lon=-3.6318, lat=40.4381\n",
      "No suitable images found for location: lon=-3.632, lat=40.438\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.6317631169605926, lat=40.43811291788501\n",
      "No suitable images found for location: lon=-3.698323, lat=40.417874\n",
      "No suitable images found for location: lon=-3.69832, lat=40.41787\n",
      "No suitable images found for location: lon=-3.6983, lat=40.4179\n",
      "No suitable images found for location: lon=-3.698, lat=40.418\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.698323418142352, lat=40.41787445445756\n",
      "No suitable images found for location: lon=2.143038, lat=41.383359\n",
      "No suitable images found for location: lon=2.14304, lat=41.38336\n",
      "No suitable images found for location: lon=2.1430, lat=41.3834\n",
      "No suitable images found for location: lon=2.143, lat=41.383\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=2.143037845248417, lat=41.38335942948304\n",
      "No suitable images found for location: lon=-2.121564, lat=40.054215\n",
      "No suitable images found for location: lon=-2.12156, lat=40.05421\n",
      "No suitable images found for location: lon=-2.1216, lat=40.0542\n",
      "No suitable images found for location: lon=-2.122, lat=40.054\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-2.121563815192116, lat=40.05421488538511\n",
      "No suitable images found for location: lon=2.160706, lat=41.394844\n",
      "No suitable images found for location: lon=2.16071, lat=41.39484\n",
      "No suitable images found for location: lon=2.1607, lat=41.3948\n",
      "No suitable images found for location: lon=2.161, lat=41.395\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=2.160705911947717, lat=41.394844454042506\n",
      "No suitable images found for location: lon=2.167804, lat=41.390444\n",
      "No suitable images found for location: lon=2.16780, lat=41.39044\n",
      "No suitable images found for location: lon=2.1678, lat=41.3904\n",
      "No suitable images found for location: lon=2.168, lat=41.390\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=2.1678037832421944, lat=41.39044407057673\n",
      "No suitable images found for location: lon=2.168559, lat=41.390100\n",
      "No suitable images found for location: lon=2.16856, lat=41.39010\n",
      "No suitable images found for location: lon=2.1686, lat=41.3901\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1936.0_12102_8b47d_57ada182da1df.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1936.0_2817074_6d3fd_5f1d181169e0f.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1936.0_3881241_60a6cabd571e9794d328746745ec3fe3.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1936.0_4549273_395e13d83f4f9b336ea00862b3c3448a.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1936.0_6530105_571eb_63b6ecfc0ba1a.jpg\n",
      "No suitable images found for location: lon=2.188949, lat=41.398876\n",
      "No suitable images found for location: lon=2.18895, lat=41.39888\n",
      "No suitable images found for location: lon=2.1889, lat=41.3989\n",
      "No suitable images found for location: lon=2.189, lat=41.399\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=2.1889493798287285, lat=41.39887608136772\n",
      "No suitable images found for location: lon=2.150699, lat=41.399341\n",
      "No suitable images found for location: lon=2.15070, lat=41.39934\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1975.0_12101_cc321_57ada101069fa.jpg. Skipping download.\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1975.0_12102_048dc_57ada19a83d38.jpg. Skipping download.\n",
      "No suitable images found for location: lon=2.155585, lat=41.394043\n",
      "No suitable images found for location: lon=2.15559, lat=41.39404\n",
      "No suitable images found for location: lon=2.1556, lat=41.3940\n",
      "No suitable images found for location: lon=2.156, lat=41.394\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=2.1555854625817545, lat=41.394042717335765\n",
      "No suitable images found for location: lon=-3.693594, lat=40.422320\n",
      "No suitable images found for location: lon=-3.69359, lat=40.42232\n",
      "No suitable images found for location: lon=-3.6936, lat=40.4223\n",
      "No suitable images found for location: lon=-3.694, lat=40.422\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.6935944662345297, lat=40.42231961362918\n",
      "No suitable images found for location: lon=-3.282000, lat=40.617255\n",
      "No suitable images found for location: lon=-3.28200, lat=40.61725\n",
      "No suitable images found for location: lon=-3.2820, lat=40.6173\n",
      "No suitable images found for location: lon=-3.282, lat=40.617\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.282000026605806, lat=40.61725458550439\n",
      "No suitable images found for location: lon=-4.422614, lat=36.717170\n",
      "No suitable images found for location: lon=-4.42261, lat=36.71717\n",
      "No suitable images found for location: lon=-4.4226, lat=36.7172\n",
      "No suitable images found for location: lon=-4.423, lat=36.717\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-4.422614383495336, lat=36.71717004360905\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1970.0_1263515_0_2180f_29.jpg. Skipping download.\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1970.0_1305415_2_c674a_1.jpg. Skipping download.\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1970.0_1582081_0_7f529_26.jpg. Skipping download.\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1970.0_2279478_3_b9b53_25.jpg. Skipping download.\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1970.0_2956154_26f81bfdf6bf86f633606d47529cfe03.jpg. Skipping download.\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1970.0_3309029_aa63e4c1bcfe447b1d421d96ae44d7cf.jpg. Skipping download.\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1970.0_3537073_147a3_60843ff2edff9.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1970.0_4194361_f1482f2b9d23a027df0728d3962c5c42.jpg\n",
      "No suitable images found for location: lon=-3.687008, lat=40.433892\n",
      "No suitable images found for location: lon=-3.68701, lat=40.43389\n",
      "No suitable images found for location: lon=-3.6870, lat=40.4339\n",
      "No suitable images found for location: lon=-3.687, lat=40.434\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.6870077929523934, lat=40.43389152295964\n",
      "No suitable images found for location: lon=-3.688059, lat=40.430907\n",
      "No suitable images found for location: lon=-3.68806, lat=40.43091\n",
      "No suitable images found for location: lon=-3.6881, lat=40.4309\n",
      "No suitable images found for location: lon=-3.688, lat=40.431\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.688058713710498, lat=40.43090742624928\n",
      "No suitable images found for location: lon=-3.719126, lat=40.386076\n",
      "No suitable images found for location: lon=-3.71913, lat=40.38608\n",
      "No suitable images found for location: lon=-3.7191, lat=40.3861\n",
      "No suitable images found for location: lon=-3.719, lat=40.386\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.7191260760455296, lat=40.386075510466384\n",
      "No suitable images found for location: lon=-3.630664, lat=40.433858\n",
      "No suitable images found for location: lon=-3.63066, lat=40.43386\n",
      "No suitable images found for location: lon=-3.6307, lat=40.4339\n",
      "No suitable images found for location: lon=-3.631, lat=40.434\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.630664457379409, lat=40.4338575557291\n",
      "No suitable images found for location: lon=-3.828115, lat=40.474248\n",
      "No suitable images found for location: lon=-3.82812, lat=40.47425\n",
      "No suitable images found for location: lon=-3.8281, lat=40.4742\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1997.0_1336181_24_7c0da_31.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1997.0_5959097_053d145c065c03983eb9b3b4084e7d11.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1997.0_8541033_a494da38dbffe220d9d834465bf0fe39.jpg\n",
      "No suitable images found for location: lon=-3.689340, lat=40.468334\n",
      "No suitable images found for location: lon=-3.68934, lat=40.46833\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1965.0_1517482_44b43_5cffc20b61e74.jpg. Skipping download.\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1965.0_3551433_9edc4_608fdb3469de6.jpg. Skipping download.\n",
      "No suitable images found for location: lon=-3.816615, lat=40.337793\n",
      "No suitable images found for location: lon=-3.81662, lat=40.33779\n",
      "No suitable images found for location: lon=-3.8166, lat=40.3378\n",
      "No suitable images found for location: lon=-3.817, lat=40.338\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.816615030325632, lat=40.33779336050903\n",
      "No suitable images found for location: lon=-3.670825, lat=40.448908\n",
      "No suitable images found for location: lon=-3.67083, lat=40.44891\n",
      "No suitable images found for location: lon=-3.6708, lat=40.4489\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1993.0_3324877_00a832bd47103dbbb0257c807a58f000.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1993.0_3479405_4fbc8321028f3bf6643fd93c74617d8d.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1993.0_3499077_88ea6_60757ddaefd5c.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1993.0_5420297_c65b884a3a5e7f5277e858e22883c0d1.jpg\n",
      "No suitable images found for location: lon=2.194837, lat=41.385807\n",
      "No suitable images found for location: lon=2.19484, lat=41.38581\n",
      "No suitable images found for location: lon=2.1948, lat=41.3858\n",
      "No suitable images found for location: lon=2.195, lat=41.386\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=2.194836912437018, lat=41.38580683258707\n",
      "No suitable images found for location: lon=-3.623775, lat=40.436405\n",
      "No suitable images found for location: lon=-3.62377, lat=40.43641\n",
      "No suitable images found for location: lon=-3.6238, lat=40.4364\n",
      "No suitable images found for location: lon=-3.624, lat=40.436\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.623774858477805, lat=40.43640524918656\n",
      "No suitable images found for location: lon=-3.720505, lat=40.421251\n",
      "No suitable images found for location: lon=-3.72051, lat=40.42125\n",
      "No suitable images found for location: lon=-3.7205, lat=40.4213\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/2004.0_2801714_71338_5f19558dd1c92.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/2004.0_2805374_04c89_5f1a3e9dcc391.jpg\n",
      "No suitable images found for location: lon=-3.717389, lat=40.445581\n",
      "No suitable images found for location: lon=-3.71739, lat=40.44558\n",
      "No suitable images found for location: lon=-3.7174, lat=40.4456\n",
      "No suitable images found for location: lon=-3.717, lat=40.446\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.717389236555902, lat=40.445581386461086\n",
      "No suitable images found for location: lon=-3.692819, lat=40.421587\n",
      "No suitable images found for location: lon=-3.69282, lat=40.42159\n",
      "No suitable images found for location: lon=-3.6928, lat=40.4216\n",
      "No suitable images found for location: lon=-3.693, lat=40.422\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.692819136843217, lat=40.42158737155827\n",
      "No suitable images found for location: lon=-3.686949, lat=40.424583\n",
      "No suitable images found for location: lon=-3.68695, lat=40.42458\n",
      "No suitable images found for location: lon=-3.6869, lat=40.4246\n",
      "No suitable images found for location: lon=-3.687, lat=40.425\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.686948785347917, lat=40.42458304517949\n",
      "No suitable images found for location: lon=-3.686043, lat=40.427376\n",
      "No suitable images found for location: lon=-3.68604, lat=40.42738\n",
      "No suitable images found for location: lon=-3.6860, lat=40.4274\n",
      "No suitable images found for location: lon=-3.686, lat=40.427\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.686042687237681, lat=40.42737639468206\n",
      "No suitable images found for location: lon=-3.690905, lat=40.422566\n",
      "No suitable images found for location: lon=-3.69090, lat=40.42257\n",
      "No suitable images found for location: lon=-3.6909, lat=40.4226\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1956.0_14592_e0a45_57bef42d47c17.jpg. Skipping download.\n",
      "File already exists: /workspace/workspace/ufo-prediction/image_data/1956.0_27505_9a9b0_582c3be024810.jpg. Skipping download.\n",
      "No suitable images found for location: lon=-3.678176, lat=40.392366\n",
      "No suitable images found for location: lon=-3.67818, lat=40.39237\n",
      "No suitable images found for location: lon=-3.6782, lat=40.3924\n",
      "No suitable images found for location: lon=-3.678, lat=40.392\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.678176104581512, lat=40.39236621478853\n",
      "No suitable images found for location: lon=-3.665503, lat=40.497859\n",
      "No suitable images found for location: lon=-3.66550, lat=40.49786\n",
      "No suitable images found for location: lon=-3.6655, lat=40.4979\n",
      "No suitable images found for location: lon=-3.666, lat=40.498\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.665502509026653, lat=40.497858621000105\n",
      "No suitable images found for location: lon=2.165636, lat=41.392142\n",
      "No suitable images found for location: lon=2.16564, lat=41.39214\n",
      "No suitable images found for location: lon=2.1656, lat=41.3921\n",
      "No suitable images found for location: lon=2.166, lat=41.392\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=2.1656360310352563, lat=41.392141610068926\n",
      "No suitable images found for location: lon=-3.601716, lat=40.450700\n",
      "No suitable images found for location: lon=-3.60172, lat=40.45070\n",
      "No suitable images found for location: lon=-3.6017, lat=40.4507\n",
      "No suitable images found for location: lon=-3.602, lat=40.451\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.6017157260103865, lat=40.45069981743553\n",
      "No suitable images found for location: lon=-3.686989, lat=40.434118\n",
      "No suitable images found for location: lon=-3.68699, lat=40.43412\n",
      "No suitable images found for location: lon=-3.6870, lat=40.4341\n",
      "No suitable images found for location: lon=-3.687, lat=40.434\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.6869894039717854, lat=40.43411845911493\n",
      "No suitable images found for location: lon=-5.841734, lat=37.339719\n",
      "No suitable images found for location: lon=-5.84173, lat=37.33972\n",
      "No suitable images found for location: lon=-5.8417, lat=37.3397\n",
      "No suitable images found for location: lon=-5.842, lat=37.340\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-5.8417339711487255, lat=37.339719365129504\n",
      "No suitable images found for location: lon=-3.690305, lat=40.428225\n",
      "No suitable images found for location: lon=-3.69030, lat=40.42823\n",
      "No suitable images found for location: lon=-3.6903, lat=40.4282\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1940.0_2791838_68b1c_5f16ee9814daf.jpg\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1940.0_6742937_a65609e8b5a74d22f25c08ab2733f240.jpg\n",
      "No suitable images found for location: lon=-3.684833, lat=40.433457\n",
      "No suitable images found for location: lon=-3.68483, lat=40.43346\n",
      "No suitable images found for location: lon=-3.6848, lat=40.4335\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/2018.0_2799522_aa346_5f18c074888e9.jpg\n",
      "No suitable images found for location: lon=2.176166, lat=41.391328\n",
      "No suitable images found for location: lon=2.17617, lat=41.39133\n",
      "No suitable images found for location: lon=2.1762, lat=41.3913\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1900.0_5026137_5c20e_6251429b34155.jpg\n",
      "No suitable images found for location: lon=-3.630664, lat=40.433858\n",
      "No suitable images found for location: lon=-3.63066, lat=40.43386\n",
      "No suitable images found for location: lon=-3.6307, lat=40.4339\n",
      "No suitable images found for location: lon=-3.631, lat=40.434\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.630664457379409, lat=40.4338575557291\n",
      "No suitable images found for location: lon=-3.838598, lat=43.427393\n",
      "No suitable images found for location: lon=-3.83860, lat=43.42739\n",
      "No suitable images found for location: lon=-3.8386, lat=43.4274\n",
      "No suitable images found for location: lon=-3.839, lat=43.427\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.838598049185864, lat=43.42739260052197\n",
      "No suitable images found for location: lon=2.732440, lat=39.632641\n",
      "No suitable images found for location: lon=2.73244, lat=39.63264\n",
      "No suitable images found for location: lon=2.7324, lat=39.6326\n",
      "No suitable images found for location: lon=2.732, lat=39.633\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=2.732440119688764, lat=39.63264082926448\n",
      "No suitable images found for location: lon=-6.211365, lat=36.616724\n",
      "No suitable images found for location: lon=-6.21136, lat=36.61672\n",
      "No suitable images found for location: lon=-6.2114, lat=36.6167\n",
      "No suitable images found for location: lon=-6.211, lat=36.617\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-6.211364506256349, lat=36.61672421054904\n",
      "No suitable images found for location: lon=-4.791468, lat=37.896091\n",
      "No suitable images found for location: lon=-4.79147, lat=37.89609\n",
      "No suitable images found for location: lon=-4.7915, lat=37.8961\n",
      "No suitable images found for location: lon=-4.791, lat=37.896\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-4.79146773540764, lat=37.89609113377536\n",
      "No suitable images found for location: lon=-4.432209, lat=36.734649\n",
      "No suitable images found for location: lon=-4.43221, lat=36.73465\n",
      "No suitable images found for location: lon=-4.4322, lat=36.7346\n",
      "No suitable images found for location: lon=-4.432, lat=36.735\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-4.432208800457712, lat=36.73464903753526\n",
      "No suitable images found for location: lon=2.053114, lat=41.482771\n",
      "No suitable images found for location: lon=2.05311, lat=41.48277\n",
      "No suitable images found for location: lon=2.0531, lat=41.4828\n",
      "No suitable images found for location: lon=2.053, lat=41.483\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=2.0531142496192665, lat=41.48277106990048\n",
      "No suitable images found for location: lon=-0.469063, lat=38.367103\n",
      "No suitable images found for location: lon=-0.46906, lat=38.36710\n",
      "No suitable images found for location: lon=-0.4691, lat=38.3671\n",
      "No suitable images found for location: lon=-0.469, lat=38.367\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-0.4690633930098014, lat=38.36710319411887\n",
      "No suitable images found for location: lon=-0.368433, lat=39.468371\n",
      "No suitable images found for location: lon=-0.36843, lat=39.46837\n",
      "No suitable images found for location: lon=-0.3684, lat=39.4684\n",
      "No suitable images found for location: lon=-0.368, lat=39.468\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-0.3684327006097273, lat=39.46837143913605\n",
      "No suitable images found for location: lon=-6.389981, lat=39.468208\n",
      "No suitable images found for location: lon=-6.38998, lat=39.46821\n",
      "No suitable images found for location: lon=-6.3900, lat=39.4682\n",
      "No suitable images found for location: lon=-6.390, lat=39.468\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-6.389980775220602, lat=39.46820754954869\n",
      "No suitable images found for location: lon=-3.650714, lat=40.450015\n",
      "No suitable images found for location: lon=-3.65071, lat=40.45001\n",
      "No suitable images found for location: lon=-3.6507, lat=40.4500\n",
      "No suitable images found for location: lon=-3.651, lat=40.450\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.650714344830843, lat=40.4500147211243\n",
      "No suitable images found for location: lon=-3.888133, lat=40.518104\n",
      "No suitable images found for location: lon=-3.88813, lat=40.51810\n",
      "No suitable images found for location: lon=-3.8881, lat=40.5181\n",
      "No suitable images found for location: lon=-3.888, lat=40.518\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.888132794008485, lat=40.51810374254467\n",
      "No suitable images found for location: lon=-3.640446, lat=40.469459\n",
      "No suitable images found for location: lon=-3.64045, lat=40.46946\n",
      "No suitable images found for location: lon=-3.6404, lat=40.4695\n",
      "Image downloaded successfully: /workspace/workspace/ufo-prediction/image_data/1993.0_3514405_441a1_607db9b1957c0.jpg\n",
      "No suitable images found for location: lon=-5.329052, lat=35.894129\n",
      "No suitable images found for location: lon=-5.32905, lat=35.89413\n",
      "No suitable images found for location: lon=-5.3291, lat=35.8941\n",
      "No suitable images found for location: lon=-5.329, lat=35.894\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-5.329051530548132, lat=35.89412890382049\n",
      "No suitable images found for location: lon=-0.813223, lat=37.820039\n",
      "No suitable images found for location: lon=-0.81322, lat=37.82004\n",
      "No suitable images found for location: lon=-0.8132, lat=37.8200\n",
      "No suitable images found for location: lon=-0.813, lat=37.820\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-0.813222648270505, lat=37.82003873192238\n",
      "No suitable images found for location: lon=-3.691382, lat=40.240244\n",
      "No suitable images found for location: lon=-3.69138, lat=40.24024\n",
      "No suitable images found for location: lon=-3.6914, lat=40.2402\n",
      "No suitable images found for location: lon=-3.691, lat=40.240\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.6913816984034447, lat=40.24024392080319\n",
      "No suitable images found for location: lon=-3.920220, lat=40.568237\n",
      "No suitable images found for location: lon=-3.92022, lat=40.56824\n",
      "No suitable images found for location: lon=-3.9202, lat=40.5682\n",
      "No suitable images found for location: lon=-3.920, lat=40.568\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.920219756589552, lat=40.56823728304962\n",
      "No suitable images found for location: lon=2.104234, lat=41.362295\n",
      "No suitable images found for location: lon=2.10423, lat=41.36229\n",
      "No suitable images found for location: lon=2.1042, lat=41.3623\n",
      "No suitable images found for location: lon=2.104, lat=41.362\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=2.1042338881427307, lat=41.36229498602597\n",
      "No suitable images found for location: lon=-3.655963, lat=40.392300\n",
      "No suitable images found for location: lon=-3.65596, lat=40.39230\n",
      "No suitable images found for location: lon=-3.6560, lat=40.3923\n",
      "No suitable images found for location: lon=-3.656, lat=40.392\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.655963200616134, lat=40.39229985018068\n",
      "No suitable images found for location: lon=-8.724449, lat=42.220811\n",
      "No suitable images found for location: lon=-8.72445, lat=42.22081\n",
      "No suitable images found for location: lon=-8.7244, lat=42.2208\n",
      "No suitable images found for location: lon=-8.724, lat=42.221\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-8.724449419170359, lat=42.2208110053289\n",
      "No suitable images found for location: lon=-1.846911, lat=38.990797\n",
      "No suitable images found for location: lon=-1.84691, lat=38.99080\n",
      "No suitable images found for location: lon=-1.8469, lat=38.9908\n",
      "No suitable images found for location: lon=-1.847, lat=38.991\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-1.846910810907593, lat=38.99079686158863\n",
      "No suitable images found for location: lon=-1.149364, lat=38.032289\n",
      "No suitable images found for location: lon=-1.14936, lat=38.03229\n",
      "No suitable images found for location: lon=-1.1494, lat=38.0323\n",
      "No suitable images found for location: lon=-1.149, lat=38.032\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-1.1493640564362104, lat=38.032288602393336\n",
      "No suitable images found for location: lon=-0.355499, lat=39.452841\n",
      "No suitable images found for location: lon=-0.35550, lat=39.45284\n",
      "No suitable images found for location: lon=-0.3555, lat=39.4528\n",
      "No suitable images found for location: lon=-0.355, lat=39.453\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-0.3554988078264734, lat=39.45284138885362\n",
      "No suitable images found for location: lon=-0.170343, lat=38.972606\n",
      "No suitable images found for location: lon=-0.17034, lat=38.97261\n",
      "No suitable images found for location: lon=-0.1703, lat=38.9726\n",
      "No suitable images found for location: lon=-0.170, lat=38.973\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-0.1703434045683163, lat=38.97260561575845\n",
      "No suitable images found for location: lon=2.228079, lat=41.456724\n",
      "No suitable images found for location: lon=2.22808, lat=41.45672\n",
      "No suitable images found for location: lon=2.2281, lat=41.4567\n",
      "No suitable images found for location: lon=2.228, lat=41.457\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=2.228078972014717, lat=41.45672384665169\n",
      "No suitable images found for location: lon=-6.028618, lat=37.392066\n",
      "No suitable images found for location: lon=-6.02862, lat=37.39207\n",
      "No suitable images found for location: lon=-6.0286, lat=37.3921\n",
      "No suitable images found for location: lon=-6.029, lat=37.392\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-6.028617996259489, lat=37.39206622051829\n",
      "No suitable images found for location: lon=-6.603647, lat=42.556535\n",
      "No suitable images found for location: lon=-6.60365, lat=42.55653\n",
      "No suitable images found for location: lon=-6.6036, lat=42.5565\n",
      "No suitable images found for location: lon=-6.604, lat=42.557\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-6.603646869975418, lat=42.5565347615265\n",
      "No suitable images found for location: lon=-4.430175, lat=36.715655\n",
      "No suitable images found for location: lon=-4.43018, lat=36.71566\n",
      "No suitable images found for location: lon=-4.4302, lat=36.7157\n",
      "No suitable images found for location: lon=-4.430, lat=36.716\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-4.430175369918901, lat=36.71565524984889\n",
      "No suitable images found for location: lon=-3.663162, lat=40.381115\n",
      "No suitable images found for location: lon=-3.66316, lat=40.38111\n",
      "No suitable images found for location: lon=-3.6632, lat=40.3811\n",
      "No suitable images found for location: lon=-3.663, lat=40.381\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.663161967366593, lat=40.381114775760246\n",
      "No suitable images found for location: lon=-3.975734, lat=39.862738\n",
      "No suitable images found for location: lon=-3.97573, lat=39.86274\n",
      "No suitable images found for location: lon=-3.9757, lat=39.8627\n",
      "No suitable images found for location: lon=-3.976, lat=39.863\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.975733650641613, lat=39.86273832108993\n",
      "No suitable images found for location: lon=-0.889934, lat=41.670676\n",
      "No suitable images found for location: lon=-0.88993, lat=41.67068\n",
      "No suitable images found for location: lon=-0.8899, lat=41.6707\n",
      "No suitable images found for location: lon=-0.890, lat=41.671\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-0.8899344330545073, lat=41.6706759951992\n",
      "No suitable images found for location: lon=-4.785896, lat=37.858490\n",
      "No suitable images found for location: lon=-4.78590, lat=37.85849\n",
      "No suitable images found for location: lon=-4.7859, lat=37.8585\n",
      "No suitable images found for location: lon=-4.786, lat=37.858\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-4.7858958964673874, lat=37.85848969373771\n",
      "No suitable images found for location: lon=-3.830931, lat=40.358236\n",
      "No suitable images found for location: lon=-3.83093, lat=40.35824\n",
      "No suitable images found for location: lon=-3.8309, lat=40.3582\n",
      "No suitable images found for location: lon=-3.831, lat=40.358\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.830930793126003, lat=40.358236061573024\n",
      "No suitable images found for location: lon=-0.494892, lat=38.342341\n",
      "No suitable images found for location: lon=-0.49489, lat=38.34234\n",
      "No suitable images found for location: lon=-0.4949, lat=38.3423\n",
      "No suitable images found for location: lon=-0.495, lat=38.342\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-0.4948923676007544, lat=38.34234104289159\n",
      "No suitable images found for location: lon=-3.710817, lat=40.415165\n",
      "No suitable images found for location: lon=-3.71082, lat=40.41516\n",
      "No suitable images found for location: lon=-3.7108, lat=40.4152\n",
      "No suitable images found for location: lon=-3.711, lat=40.415\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.710817161553617, lat=40.41516496708134\n",
      "No suitable images found for location: lon=-2.461943, lat=36.840253\n",
      "No suitable images found for location: lon=-2.46194, lat=36.84025\n",
      "No suitable images found for location: lon=-2.4619, lat=36.8403\n",
      "No suitable images found for location: lon=-2.462, lat=36.840\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-2.4619431088560866, lat=36.84025345085567\n",
      "No suitable images found for location: lon=-4.667069, lat=36.553247\n",
      "No suitable images found for location: lon=-4.66707, lat=36.55325\n",
      "No suitable images found for location: lon=-4.6671, lat=36.5532\n",
      "No suitable images found for location: lon=-4.667, lat=36.553\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-4.66706918529366, lat=36.55324736696102\n",
      "No suitable images found for location: lon=-3.714556, lat=40.430154\n",
      "No suitable images found for location: lon=-3.71456, lat=40.43015\n",
      "No suitable images found for location: lon=-3.7146, lat=40.4302\n",
      "No suitable images found for location: lon=-3.715, lat=40.430\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.7145560509119178, lat=40.430154496359805\n",
      "No suitable images found for location: lon=-0.881502, lat=41.652495\n",
      "No suitable images found for location: lon=-0.88150, lat=41.65249\n",
      "No suitable images found for location: lon=-0.8815, lat=41.6525\n",
      "No suitable images found for location: lon=-0.882, lat=41.652\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-0.8815018691571542, lat=41.65249470669832\n",
      "No suitable images found for location: lon=-3.490554, lat=40.388140\n",
      "No suitable images found for location: lon=-3.49055, lat=40.38814\n",
      "No suitable images found for location: lon=-3.4906, lat=40.3881\n",
      "No suitable images found for location: lon=-3.491, lat=40.388\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.490554297221222, lat=40.388139764719845\n",
      "No suitable images found for location: lon=-3.718784, lat=40.600746\n",
      "No suitable images found for location: lon=-3.71878, lat=40.60075\n",
      "No suitable images found for location: lon=-3.7188, lat=40.6007\n",
      "No suitable images found for location: lon=-3.719, lat=40.601\n",
      "Unable to retrieve data from the API with sufficient precision for location: lon=-3.71878448706024, lat=40.60074629138333\n",
      "No suitable images found for location: lon=-3.714320, lat=40.427130\n",
      "No suitable images found for location: lon=-3.71432, lat=40.42713\n",
      "No suitable images found for location: lon=-3.7143, lat=40.4271\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [21], line 14\u001b[0m\n\u001b[1;32m     11\u001b[0m url \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttps://api.openstreetcam.org/2.0/photo/?lat=\u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m&lng=\u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(lat, lon)\n\u001b[1;32m     13\u001b[0m \u001b[38;5;66;03m# Send a GET request to the API\u001b[39;00m\n\u001b[0;32m---> 14\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     16\u001b[0m \u001b[38;5;66;03m# Check if the request was successful\u001b[39;00m\n\u001b[1;32m     17\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m response\u001b[38;5;241m.\u001b[39mstatus_code \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m200\u001b[39m:\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/requests/api.py:75\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     64\u001b[0m \u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     65\u001b[0m \n\u001b[1;32m     66\u001b[0m \u001b[38;5;124;03m:param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;124;03m:rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     72\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m     74\u001b[0m kwargs\u001b[38;5;241m.\u001b[39msetdefault(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m---> 75\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/requests/api.py:60\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     56\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     59\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 60\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/requests/sessions.py:533\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    528\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    529\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m'\u001b[39m: timeout,\n\u001b[1;32m    530\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m'\u001b[39m: allow_redirects,\n\u001b[1;32m    531\u001b[0m }\n\u001b[1;32m    532\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 533\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    535\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/requests/sessions.py:646\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    643\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    645\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 646\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    648\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    649\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/requests/adapters.py:439\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    437\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    438\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m chunked:\n\u001b[0;32m--> 439\u001b[0m         resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    440\u001b[0m \u001b[43m            \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    441\u001b[0m \u001b[43m            \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    442\u001b[0m \u001b[43m            \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    443\u001b[0m \u001b[43m            \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    444\u001b[0m \u001b[43m            \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    445\u001b[0m \u001b[43m            \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    446\u001b[0m \u001b[43m            \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    447\u001b[0m \u001b[43m            \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    448\u001b[0m \u001b[43m            \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    449\u001b[0m \u001b[43m            \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\n\u001b[1;32m    450\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    452\u001b[0m     \u001b[38;5;66;03m# Send the request.\u001b[39;00m\n\u001b[1;32m    453\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    454\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(conn, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mproxy_pool\u001b[39m\u001b[38;5;124m'\u001b[39m):\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/urllib3/connectionpool.py:665\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    662\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_prepare_proxy(conn)\n\u001b[1;32m    664\u001b[0m \u001b[38;5;66;03m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 665\u001b[0m httplib_response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    666\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    667\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    668\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    669\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    670\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    671\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    672\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    673\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    675\u001b[0m \u001b[38;5;66;03m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    676\u001b[0m \u001b[38;5;66;03m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    677\u001b[0m \u001b[38;5;66;03m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    678\u001b[0m \u001b[38;5;66;03m# mess.\u001b[39;00m\n\u001b[1;32m    679\u001b[0m response_conn \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m release_conn \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/urllib3/connectionpool.py:376\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    374\u001b[0m \u001b[38;5;66;03m# Trigger any extra validation we need to do.\u001b[39;00m\n\u001b[1;32m    375\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 376\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    377\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (SocketTimeout, BaseSSLError) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    378\u001b[0m     \u001b[38;5;66;03m# Py2 raises this as a BaseSSLError, Py3 raises it as socket timeout.\u001b[39;00m\n\u001b[1;32m    379\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_raise_timeout(err\u001b[38;5;241m=\u001b[39me, url\u001b[38;5;241m=\u001b[39murl, timeout_value\u001b[38;5;241m=\u001b[39mconn\u001b[38;5;241m.\u001b[39mtimeout)\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/urllib3/connectionpool.py:996\u001b[0m, in \u001b[0;36mHTTPSConnectionPool._validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m    994\u001b[0m \u001b[38;5;66;03m# Force connect early to allow us to validate the connection.\u001b[39;00m\n\u001b[1;32m    995\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(conn, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msock\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):  \u001b[38;5;66;03m# AppEngine might not have  `.sock`\u001b[39;00m\n\u001b[0;32m--> 996\u001b[0m     \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    998\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m conn\u001b[38;5;241m.\u001b[39mis_verified:\n\u001b[1;32m    999\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m   1000\u001b[0m         (\n\u001b[1;32m   1001\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnverified HTTPS request is being made to host \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1006\u001b[0m         InsecureRequestWarning,\n\u001b[1;32m   1007\u001b[0m     )\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/urllib3/connection.py:366\u001b[0m, in \u001b[0;36mVerifiedHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    358\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m    359\u001b[0m     \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mca_certs\n\u001b[1;32m    360\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mca_cert_dir\n\u001b[1;32m    361\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m default_ssl_context\n\u001b[1;32m    362\u001b[0m     \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(context, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mload_default_certs\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[1;32m    363\u001b[0m ):\n\u001b[1;32m    364\u001b[0m     context\u001b[38;5;241m.\u001b[39mload_default_certs()\n\u001b[0;32m--> 366\u001b[0m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock \u001b[38;5;241m=\u001b[39m \u001b[43mssl_wrap_socket\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    367\u001b[0m \u001b[43m    \u001b[49m\u001b[43msock\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    368\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkeyfile\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkey_file\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    369\u001b[0m \u001b[43m    \u001b[49m\u001b[43mcertfile\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcert_file\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    370\u001b[0m \u001b[43m    \u001b[49m\u001b[43mkey_password\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mkey_password\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    371\u001b[0m \u001b[43m    \u001b[49m\u001b[43mca_certs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mca_certs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    372\u001b[0m \u001b[43m    \u001b[49m\u001b[43mca_cert_dir\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mca_cert_dir\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    373\u001b[0m \u001b[43m    \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    374\u001b[0m \u001b[43m    \u001b[49m\u001b[43mssl_context\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mcontext\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    375\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    377\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39massert_fingerprint:\n\u001b[1;32m    378\u001b[0m     assert_fingerprint(\n\u001b[1;32m    379\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msock\u001b[38;5;241m.\u001b[39mgetpeercert(binary_form\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m), \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39massert_fingerprint\n\u001b[1;32m    380\u001b[0m     )\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/urllib3/util/ssl_.py:370\u001b[0m, in \u001b[0;36mssl_wrap_socket\u001b[0;34m(sock, keyfile, certfile, cert_reqs, ca_certs, server_hostname, ssl_version, ciphers, ssl_context, ca_cert_dir, key_password)\u001b[0m\n\u001b[1;32m    366\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m (\n\u001b[1;32m    367\u001b[0m     server_hostname \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m \u001b[38;5;129;01mand\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m is_ipaddress(server_hostname)\n\u001b[1;32m    368\u001b[0m ) \u001b[38;5;129;01mor\u001b[39;00m IS_SECURETRANSPORT:\n\u001b[1;32m    369\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m HAS_SNI \u001b[38;5;129;01mand\u001b[39;00m server_hostname \u001b[38;5;129;01mis\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m:\n\u001b[0;32m--> 370\u001b[0m         \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mcontext\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mwrap_socket\u001b[49m\u001b[43m(\u001b[49m\u001b[43msock\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    372\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m    373\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mAn HTTPS request has been made, but the SNI (Server Name \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    374\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mIndication) extension to TLS is not available on this platform. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m    380\u001b[0m         SNIMissingWarning,\n\u001b[1;32m    381\u001b[0m     )\n\u001b[1;32m    383\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m context\u001b[38;5;241m.\u001b[39mwrap_socket(sock)\n",
      "File \u001b[0;32m~usr/lib/python3.8/ssl.py:500\u001b[0m, in \u001b[0;36mSSLContext.wrap_socket\u001b[0;34m(self, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, session)\u001b[0m\n\u001b[1;32m    494\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mwrap_socket\u001b[39m(\u001b[38;5;28mself\u001b[39m, sock, server_side\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mFalse\u001b[39;00m,\n\u001b[1;32m    495\u001b[0m                 do_handshake_on_connect\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m    496\u001b[0m                 suppress_ragged_eofs\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mTrue\u001b[39;00m,\n\u001b[1;32m    497\u001b[0m                 server_hostname\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m, session\u001b[38;5;241m=\u001b[39m\u001b[38;5;28;01mNone\u001b[39;00m):\n\u001b[1;32m    498\u001b[0m     \u001b[38;5;66;03m# SSLSocket class handles server_hostname encoding before it calls\u001b[39;00m\n\u001b[1;32m    499\u001b[0m     \u001b[38;5;66;03m# ctx._wrap_socket()\u001b[39;00m\n\u001b[0;32m--> 500\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msslsocket_class\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_create\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    501\u001b[0m \u001b[43m        \u001b[49m\u001b[43msock\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msock\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    502\u001b[0m \u001b[43m        \u001b[49m\u001b[43mserver_side\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_side\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    503\u001b[0m \u001b[43m        \u001b[49m\u001b[43mdo_handshake_on_connect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mdo_handshake_on_connect\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    504\u001b[0m \u001b[43m        \u001b[49m\u001b[43msuppress_ragged_eofs\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msuppress_ragged_eofs\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    505\u001b[0m \u001b[43m        \u001b[49m\u001b[43mserver_hostname\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mserver_hostname\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    506\u001b[0m \u001b[43m        \u001b[49m\u001b[43mcontext\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[43m,\u001b[49m\n\u001b[1;32m    507\u001b[0m \u001b[43m        \u001b[49m\u001b[43msession\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43msession\u001b[49m\n\u001b[1;32m    508\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~usr/lib/python3.8/ssl.py:1040\u001b[0m, in \u001b[0;36mSSLSocket._create\u001b[0;34m(cls, sock, server_side, do_handshake_on_connect, suppress_ragged_eofs, server_hostname, context, session)\u001b[0m\n\u001b[1;32m   1037\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0.0\u001b[39m:\n\u001b[1;32m   1038\u001b[0m             \u001b[38;5;66;03m# non-blocking\u001b[39;00m\n\u001b[1;32m   1039\u001b[0m             \u001b[38;5;28;01mraise\u001b[39;00m \u001b[38;5;167;01mValueError\u001b[39;00m(\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mdo_handshake_on_connect should not be specified for non-blocking sockets\u001b[39m\u001b[38;5;124m\"\u001b[39m)\n\u001b[0;32m-> 1040\u001b[0m         \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdo_handshake\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1041\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (\u001b[38;5;167;01mOSError\u001b[39;00m, \u001b[38;5;167;01mValueError\u001b[39;00m):\n\u001b[1;32m   1042\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mclose()\n",
      "File \u001b[0;32m~usr/lib/python3.8/ssl.py:1309\u001b[0m, in \u001b[0;36mSSLSocket.do_handshake\u001b[0;34m(self, block)\u001b[0m\n\u001b[1;32m   1307\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m timeout \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m0.0\u001b[39m \u001b[38;5;129;01mand\u001b[39;00m block:\n\u001b[1;32m   1308\u001b[0m         \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msettimeout(\u001b[38;5;28;01mNone\u001b[39;00m)\n\u001b[0;32m-> 1309\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_sslobj\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mdo_handshake\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m   1310\u001b[0m \u001b[38;5;28;01mfinally\u001b[39;00m:\n\u001b[1;32m   1311\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msettimeout(timeout)\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "for index, row in kartaview_keys.iterrows():\n",
    "    precision = 6  # Start with 6 decimal places\n",
    "    success = False  # Flag to indicate if the request was successful\n",
    "\n",
    "    while precision > 2 and not success:\n",
    "        # Format lon and lat to the current precision\n",
    "        lon = f\"{row['lon']:.{precision}f}\"\n",
    "        lat = f\"{row['lat']:.{precision}f}\"\n",
    "\n",
    "        # Construct the API URL\n",
    "        url = \"https://api.openstreetcam.org/2.0/photo/?lat={}&lng={}\".format(lat, lon)\n",
    "\n",
    "        # Send a GET request to the API\n",
    "        response = requests.get(url)\n",
    "\n",
    "        # Check if the request was successful\n",
    "        if response.status_code == 200:\n",
    "            data = response.json()\n",
    "            url_pattern = r'https://storage\\d+\\.openstreetcam\\.org/files/photo/\\d+/\\d+/\\d+/[^\"]+\\.jpg'\n",
    "            urls = re.findall(url_pattern, json.dumps(data))\n",
    "\n",
    "            filtered_urls = [\n",
    "                url for url in urls\n",
    "                if not any(x in url.rsplit('/', 2)[-2] for x in [\"{{sizeprefix}}\", \"proc\"]) and\n",
    "                (\"th\" in url.rsplit('/', 2)[-2] and not \"lth\" in url.rsplit('/', 2)[-2])\n",
    "            ]\n",
    "\n",
    "            if filtered_urls:\n",
    "                for image_url in filtered_urls:\n",
    "                    file_name = \"{}_{}\".format(row['age_right'], image_url.split('/')[-1])\n",
    "                    if not file_name.endswith('.jpg'):\n",
    "                        file_name += '.jpg'\n",
    "                    file_path = os.path.join(current_directory, file_name)\n",
    "\n",
    "                    # Check if the file already exists\n",
    "                    if os.path.exists(file_path):\n",
    "                        print(f\"File already exists: {file_path}. Skipping download.\")\n",
    "                    else:\n",
    "                        image_response = requests.get(image_url)\n",
    "\n",
    "                        if image_response.status_code == 200:\n",
    "                            with open(file_path, 'wb') as f:\n",
    "                                f.write(image_response.content)\n",
    "                            print(\"Image downloaded successfully: {}\".format(file_path))\n",
    "                        else:\n",
    "                            print(\"Failed to download the image.\")\n",
    "                success = True  # Mark success as True to exit the while loop\n",
    "            else:\n",
    "                print(\"No suitable images found for location: lon={}, lat={}\".format(lon, lat))\n",
    "                precision -= 1  # Reduce precision by one decimal place\n",
    "        else:\n",
    "            print(\"Failed to retrieve data from the API for location: lon={}, lat={}. Trying with reduced precision.\")\n",
    "            \n",
    "\n",
    "    # After processing, remove the row from df_subset\n",
    "    kartaview_keys = kartaview_keys.drop(index)\n",
    "\n",
    "    # Save the updated DataFrame to a CSV file\n",
    "    kartaview_keys.to_csv(processed_df_path, index=False)\n",
    "\n",
    "    if not success:\n",
    "        print(\"Unable to retrieve data from the API with sufficient precision for location: lon={}, lat={}\".format(row['lon'], row['lat']))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Failed to retrieve data from the API for location: lon=4.519975, lat=52.075455\n",
      "Failed to retrieve data from the API for location: lon=5.970008, lat=50.885978\n",
      "Failed to retrieve data from the API for location: lon=5.620564, lat=52.349998\n",
      "Failed to retrieve data from the API for location: lon=4.540407, lat=52.060026\n",
      "Failed to retrieve data from the API for location: lon=5.619979, lat=52.026556\n",
      "Failed to retrieve data from the API for location: lon=4.700046, lat=52.295673\n",
      "Failed to retrieve data from the API for location: lon=5.978112, lat=51.979975\n",
      "Failed to retrieve data from the API for location: lon=5.07596, lat=52.070016\n",
      "Failed to retrieve data from the API for location: lon=4.880037, lat=52.3599\n"
     ]
    },
    {
     "ename": "KeyboardInterrupt",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m                         Traceback (most recent call last)",
      "Cell \u001b[0;32mIn [14], line 18\u001b[0m\n\u001b[1;32m     13\u001b[0m url \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mhttps://api.openstreetcam.org/2.0/photo/?lat=\u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m&lng=\u001b[39m\u001b[38;5;132;01m{}\u001b[39;00m\u001b[38;5;124m\"\u001b[39m\u001b[38;5;241m.\u001b[39mformat(lat, lon)\n\u001b[1;32m     14\u001b[0m \u001b[38;5;66;03m#url=\"http://api.openstreetcam.org/2.0/list/nearby-photos/?lat={}&lng={}\".format(lat, lon)\u001b[39;00m\n\u001b[1;32m     15\u001b[0m \u001b[38;5;66;03m#print(url)\u001b[39;00m\n\u001b[1;32m     16\u001b[0m \n\u001b[1;32m     17\u001b[0m \u001b[38;5;66;03m# Send a GET request to the API\u001b[39;00m\n\u001b[0;32m---> 18\u001b[0m response \u001b[38;5;241m=\u001b[39m \u001b[43mrequests\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mget\u001b[49m\u001b[43m(\u001b[49m\u001b[43murl\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     20\u001b[0m \u001b[38;5;66;03m# Proceed if the request was successful\u001b[39;00m\n\u001b[1;32m     21\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m response\u001b[38;5;241m.\u001b[39mstatus_code \u001b[38;5;241m==\u001b[39m \u001b[38;5;241m200\u001b[39m:\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/requests/api.py:75\u001b[0m, in \u001b[0;36mget\u001b[0;34m(url, params, **kwargs)\u001b[0m\n\u001b[1;32m     64\u001b[0m \u001b[38;5;124mr\u001b[39m\u001b[38;5;124;03m\"\"\"Sends a GET request.\u001b[39;00m\n\u001b[1;32m     65\u001b[0m \n\u001b[1;32m     66\u001b[0m \u001b[38;5;124;03m:param url: URL for the new :class:`Request` object.\u001b[39;00m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m     71\u001b[0m \u001b[38;5;124;03m:rtype: requests.Response\u001b[39;00m\n\u001b[1;32m     72\u001b[0m \u001b[38;5;124;03m\"\"\"\u001b[39;00m\n\u001b[1;32m     74\u001b[0m kwargs\u001b[38;5;241m.\u001b[39msetdefault(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;28;01mTrue\u001b[39;00m)\n\u001b[0;32m---> 75\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mget\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43mparams\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mparams\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/requests/api.py:60\u001b[0m, in \u001b[0;36mrequest\u001b[0;34m(method, url, **kwargs)\u001b[0m\n\u001b[1;32m     56\u001b[0m \u001b[38;5;66;03m# By using the 'with' statement we are sure the session is closed, thus we\u001b[39;00m\n\u001b[1;32m     57\u001b[0m \u001b[38;5;66;03m# avoid leaving sockets open which can trigger a ResourceWarning in some\u001b[39;00m\n\u001b[1;32m     58\u001b[0m \u001b[38;5;66;03m# cases, and look like a memory leak in others.\u001b[39;00m\n\u001b[1;32m     59\u001b[0m \u001b[38;5;28;01mwith\u001b[39;00m sessions\u001b[38;5;241m.\u001b[39mSession() \u001b[38;5;28;01mas\u001b[39;00m session:\n\u001b[0;32m---> 60\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m \u001b[43msession\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[43m(\u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/requests/sessions.py:533\u001b[0m, in \u001b[0;36mSession.request\u001b[0;34m(self, method, url, params, data, headers, cookies, files, auth, timeout, allow_redirects, proxies, hooks, stream, verify, cert, json)\u001b[0m\n\u001b[1;32m    528\u001b[0m send_kwargs \u001b[38;5;241m=\u001b[39m {\n\u001b[1;32m    529\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mtimeout\u001b[39m\u001b[38;5;124m'\u001b[39m: timeout,\n\u001b[1;32m    530\u001b[0m     \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mallow_redirects\u001b[39m\u001b[38;5;124m'\u001b[39m: allow_redirects,\n\u001b[1;32m    531\u001b[0m }\n\u001b[1;32m    532\u001b[0m send_kwargs\u001b[38;5;241m.\u001b[39mupdate(settings)\n\u001b[0;32m--> 533\u001b[0m resp \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mprep\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43msend_kwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    535\u001b[0m \u001b[38;5;28;01mreturn\u001b[39;00m resp\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/requests/sessions.py:646\u001b[0m, in \u001b[0;36mSession.send\u001b[0;34m(self, request, **kwargs)\u001b[0m\n\u001b[1;32m    643\u001b[0m start \u001b[38;5;241m=\u001b[39m preferred_clock()\n\u001b[1;32m    645\u001b[0m \u001b[38;5;66;03m# Send the request\u001b[39;00m\n\u001b[0;32m--> 646\u001b[0m r \u001b[38;5;241m=\u001b[39m \u001b[43madapter\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43msend\u001b[49m\u001b[43m(\u001b[49m\u001b[43mrequest\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mkwargs\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    648\u001b[0m \u001b[38;5;66;03m# Total elapsed time of the request (approximately)\u001b[39;00m\n\u001b[1;32m    649\u001b[0m elapsed \u001b[38;5;241m=\u001b[39m preferred_clock() \u001b[38;5;241m-\u001b[39m start\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/requests/adapters.py:439\u001b[0m, in \u001b[0;36mHTTPAdapter.send\u001b[0;34m(self, request, stream, timeout, verify, cert, proxies)\u001b[0m\n\u001b[1;32m    437\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[1;32m    438\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m chunked:\n\u001b[0;32m--> 439\u001b[0m         resp \u001b[38;5;241m=\u001b[39m \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43murlopen\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    440\u001b[0m \u001b[43m            \u001b[49m\u001b[43mmethod\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    441\u001b[0m \u001b[43m            \u001b[49m\u001b[43murl\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    442\u001b[0m \u001b[43m            \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    443\u001b[0m \u001b[43m            \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mrequest\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    444\u001b[0m \u001b[43m            \u001b[49m\u001b[43mredirect\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    445\u001b[0m \u001b[43m            \u001b[49m\u001b[43massert_same_host\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    446\u001b[0m \u001b[43m            \u001b[49m\u001b[43mpreload_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    447\u001b[0m \u001b[43m            \u001b[49m\u001b[43mdecode_content\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43;01mFalse\u001b[39;49;00m\u001b[43m,\u001b[49m\n\u001b[1;32m    448\u001b[0m \u001b[43m            \u001b[49m\u001b[43mretries\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mmax_retries\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    449\u001b[0m \u001b[43m            \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout\u001b[49m\n\u001b[1;32m    450\u001b[0m \u001b[43m        \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    452\u001b[0m     \u001b[38;5;66;03m# Send the request.\u001b[39;00m\n\u001b[1;32m    453\u001b[0m     \u001b[38;5;28;01melse\u001b[39;00m:\n\u001b[1;32m    454\u001b[0m         \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;28mhasattr\u001b[39m(conn, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mproxy_pool\u001b[39m\u001b[38;5;124m'\u001b[39m):\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/urllib3/connectionpool.py:665\u001b[0m, in \u001b[0;36mHTTPConnectionPool.urlopen\u001b[0;34m(self, method, url, body, headers, retries, redirect, assert_same_host, timeout, pool_timeout, release_conn, chunked, body_pos, **response_kw)\u001b[0m\n\u001b[1;32m    662\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_prepare_proxy(conn)\n\u001b[1;32m    664\u001b[0m \u001b[38;5;66;03m# Make the request on the httplib connection object.\u001b[39;00m\n\u001b[0;32m--> 665\u001b[0m httplib_response \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_make_request\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    666\u001b[0m \u001b[43m    \u001b[49m\u001b[43mconn\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    667\u001b[0m \u001b[43m    \u001b[49m\u001b[43mmethod\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    668\u001b[0m \u001b[43m    \u001b[49m\u001b[43murl\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    669\u001b[0m \u001b[43m    \u001b[49m\u001b[43mtimeout\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mtimeout_obj\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    670\u001b[0m \u001b[43m    \u001b[49m\u001b[43mbody\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mbody\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    671\u001b[0m \u001b[43m    \u001b[49m\u001b[43mheaders\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mheaders\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    672\u001b[0m \u001b[43m    \u001b[49m\u001b[43mchunked\u001b[49m\u001b[38;5;241;43m=\u001b[39;49m\u001b[43mchunked\u001b[49m\u001b[43m,\u001b[49m\n\u001b[1;32m    673\u001b[0m \u001b[43m\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    675\u001b[0m \u001b[38;5;66;03m# If we're going to release the connection in ``finally:``, then\u001b[39;00m\n\u001b[1;32m    676\u001b[0m \u001b[38;5;66;03m# the response doesn't need to know about the connection. Otherwise\u001b[39;00m\n\u001b[1;32m    677\u001b[0m \u001b[38;5;66;03m# it will also try to release it and we'll have a double-release\u001b[39;00m\n\u001b[1;32m    678\u001b[0m \u001b[38;5;66;03m# mess.\u001b[39;00m\n\u001b[1;32m    679\u001b[0m response_conn \u001b[38;5;241m=\u001b[39m conn \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m release_conn \u001b[38;5;28;01melse\u001b[39;00m \u001b[38;5;28;01mNone\u001b[39;00m\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/urllib3/connectionpool.py:376\u001b[0m, in \u001b[0;36mHTTPConnectionPool._make_request\u001b[0;34m(self, conn, method, url, timeout, chunked, **httplib_request_kw)\u001b[0m\n\u001b[1;32m    374\u001b[0m \u001b[38;5;66;03m# Trigger any extra validation we need to do.\u001b[39;00m\n\u001b[1;32m    375\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 376\u001b[0m     \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_validate_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43mconn\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    377\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m (SocketTimeout, BaseSSLError) \u001b[38;5;28;01mas\u001b[39;00m e:\n\u001b[1;32m    378\u001b[0m     \u001b[38;5;66;03m# Py2 raises this as a BaseSSLError, Py3 raises it as socket timeout.\u001b[39;00m\n\u001b[1;32m    379\u001b[0m     \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39m_raise_timeout(err\u001b[38;5;241m=\u001b[39me, url\u001b[38;5;241m=\u001b[39murl, timeout_value\u001b[38;5;241m=\u001b[39mconn\u001b[38;5;241m.\u001b[39mtimeout)\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/urllib3/connectionpool.py:996\u001b[0m, in \u001b[0;36mHTTPSConnectionPool._validate_conn\u001b[0;34m(self, conn)\u001b[0m\n\u001b[1;32m    994\u001b[0m \u001b[38;5;66;03m# Force connect early to allow us to validate the connection.\u001b[39;00m\n\u001b[1;32m    995\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m \u001b[38;5;28mgetattr\u001b[39m(conn, \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msock\u001b[39m\u001b[38;5;124m\"\u001b[39m, \u001b[38;5;28;01mNone\u001b[39;00m):  \u001b[38;5;66;03m# AppEngine might not have  `.sock`\u001b[39;00m\n\u001b[0;32m--> 996\u001b[0m     \u001b[43mconn\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    998\u001b[0m \u001b[38;5;28;01mif\u001b[39;00m \u001b[38;5;129;01mnot\u001b[39;00m conn\u001b[38;5;241m.\u001b[39mis_verified:\n\u001b[1;32m    999\u001b[0m     warnings\u001b[38;5;241m.\u001b[39mwarn(\n\u001b[1;32m   1000\u001b[0m         (\n\u001b[1;32m   1001\u001b[0m             \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mUnverified HTTPS request is being made to host \u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124m. \u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[0;32m   (...)\u001b[0m\n\u001b[1;32m   1006\u001b[0m         InsecureRequestWarning,\n\u001b[1;32m   1007\u001b[0m     )\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/urllib3/connection.py:314\u001b[0m, in \u001b[0;36mVerifiedHTTPSConnection.connect\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    312\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mconnect\u001b[39m(\u001b[38;5;28mself\u001b[39m):\n\u001b[1;32m    313\u001b[0m     \u001b[38;5;66;03m# Add certificate verification\u001b[39;00m\n\u001b[0;32m--> 314\u001b[0m     conn \u001b[38;5;241m=\u001b[39m \u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_new_conn\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    315\u001b[0m     hostname \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhost\n\u001b[1;32m    317\u001b[0m     \u001b[38;5;66;03m# Google App Engine's httplib does not define _tunnel_host\u001b[39;00m\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/urllib3/connection.py:159\u001b[0m, in \u001b[0;36mHTTPConnection._new_conn\u001b[0;34m(self)\u001b[0m\n\u001b[1;32m    156\u001b[0m     extra_kw[\u001b[38;5;124m\"\u001b[39m\u001b[38;5;124msocket_options\u001b[39m\u001b[38;5;124m\"\u001b[39m] \u001b[38;5;241m=\u001b[39m \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39msocket_options\n\u001b[1;32m    158\u001b[0m \u001b[38;5;28;01mtry\u001b[39;00m:\n\u001b[0;32m--> 159\u001b[0m     conn \u001b[38;5;241m=\u001b[39m \u001b[43mconnection\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mcreate_connection\u001b[49m\u001b[43m(\u001b[49m\n\u001b[1;32m    160\u001b[0m \u001b[43m        \u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43m_dns_host\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mport\u001b[49m\u001b[43m)\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;28;43mself\u001b[39;49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mtimeout\u001b[49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[38;5;241;43m*\u001b[39;49m\u001b[43mextra_kw\u001b[49m\n\u001b[1;32m    161\u001b[0m \u001b[43m    \u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m    163\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m SocketTimeout:\n\u001b[1;32m    164\u001b[0m     \u001b[38;5;28;01mraise\u001b[39;00m ConnectTimeoutError(\n\u001b[1;32m    165\u001b[0m         \u001b[38;5;28mself\u001b[39m,\n\u001b[1;32m    166\u001b[0m         \u001b[38;5;124m\"\u001b[39m\u001b[38;5;124mConnection to \u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m timed out. (connect timeout=\u001b[39m\u001b[38;5;132;01m%s\u001b[39;00m\u001b[38;5;124m)\u001b[39m\u001b[38;5;124m\"\u001b[39m\n\u001b[1;32m    167\u001b[0m         \u001b[38;5;241m%\u001b[39m (\u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mhost, \u001b[38;5;28mself\u001b[39m\u001b[38;5;241m.\u001b[39mtimeout),\n\u001b[1;32m    168\u001b[0m     )\n",
      "File \u001b[0;32m~usr/lib/python3/dist-packages/urllib3/util/connection.py:74\u001b[0m, in \u001b[0;36mcreate_connection\u001b[0;34m(address, timeout, source_address, socket_options)\u001b[0m\n\u001b[1;32m     72\u001b[0m     \u001b[38;5;28;01mif\u001b[39;00m source_address:\n\u001b[1;32m     73\u001b[0m         sock\u001b[38;5;241m.\u001b[39mbind(source_address)\n\u001b[0;32m---> 74\u001b[0m     \u001b[43msock\u001b[49m\u001b[38;5;241;43m.\u001b[39;49m\u001b[43mconnect\u001b[49m\u001b[43m(\u001b[49m\u001b[43msa\u001b[49m\u001b[43m)\u001b[49m\n\u001b[1;32m     75\u001b[0m     \u001b[38;5;28;01mreturn\u001b[39;00m sock\n\u001b[1;32m     77\u001b[0m \u001b[38;5;28;01mexcept\u001b[39;00m socket\u001b[38;5;241m.\u001b[39merror \u001b[38;5;28;01mas\u001b[39;00m e:\n",
      "\u001b[0;31mKeyboardInterrupt\u001b[0m: "
     ]
    }
   ],
   "source": [
    "# Iterate over each row in df_subset\n",
    "for index, row in kartaview_keys.iterrows():\n",
    "    # Format lon and lat to ensure they are strings with up to 8 decimal places\n",
    "    lon = \"{:.6f}\".format(row['lon']).rstrip('0').rstrip('.')\n",
    "    lat = \"{:.6f}\".format(row['lat']).rstrip('0').rstrip('.')\n",
    "\n",
    "    # Check if lon and lat have less than 8 digits after the decimal\n",
    "    if len(lon.split('.')[-1]) < 3 or len(lat.split('.')[-1]) < 3:\n",
    "        print(\"* Location not specific enough\")\n",
    "        continue\n",
    "\n",
    "    # Construct the API URL\n",
    "    url = \"https://api.openstreetcam.org/2.0/photo/?lat={}&lng={}\".format(lat, lon)\n",
    "    #url=\"http://api.openstreetcam.org/2.0/list/nearby-photos/?lat={}&lng={}\".format(lat, lon)\n",
    "    #print(url)\n",
    "\n",
    "    # Send a GET request to the API\n",
    "    response = requests.get(url)\n",
    "\n",
    "    # Proceed if the request was successful\n",
    "    if response.status_code == 200:\n",
    "        data = response.json()\n",
    "\n",
    "        # Define a regular expression pattern to match the URLs\n",
    "        url_pattern = r'https://storage\\d+\\.openstreetcam\\.org/files/photo/\\d+/\\d+/\\d+/[^\"]+\\.jpg'\n",
    "\n",
    "        # Find all matches of the pattern\n",
    "        urls = re.findall(url_pattern, json.dumps(data))\n",
    "\n",
    "        # Filter URLs based on the specified conditions\n",
    "        filtered_urls = [url for url in urls if not any(x in url.rsplit('/', 2)[-2] for x in [\"{{sizeprefix}}\", \"proc\"]) and (\"th\" in url.rsplit('/', 2)[-2] and not \"lth\" in url.rsplit('/', 2)[-2])]\n",
    "\n",
    "        # Download images from filtered URLs\n",
    "        for image_url in filtered_urls:\n",
    "            file_name = image_url.split('/')[-1]\n",
    "            if not file_name.endswith('.jpg'):\n",
    "                file_name += '.jpg'\n",
    "\n",
    "            # Name the .jpg using age_right followed by the original file_name\n",
    "            file_name = \"{}_{}\".format(row['age_right'], file_name)\n",
    "\n",
    "            # Combine the directory with the filename\n",
    "            file_path = os.path.join(current_directory, file_name)\n",
    "\n",
    "            # Send a GET request to fetch the image\n",
    "            image_response = requests.get(image_url)\n",
    "\n",
    "            if image_response.status_code == 200:\n",
    "                # Save the image to a file\n",
    "                with open(file_path, 'wb') as f:\n",
    "                    f.write(image_response.content)\n",
    "                print(\"Image downloaded successfully: {}\".format(file_path))\n",
    "            else:\n",
    "                print(\"Failed to download the image.\")\n",
    "        else:\n",
    "            print(\"Failed to retrieve data from the API for location: lon={}, lat={}\".format(lon, lat))\n",
    "        # After processing, remove the row from df_subset\n",
    "    kartaview_keys = kartaview_keys.drop(index)\n",
    "\n",
    "    # Save the updated DataFrame to a CSV file\n",
    "    kartaview_keys.to_csv(processed_df_path, index=False)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['https://storage2.openstreetcam.org/files/photo/2016/8/25/th/14682_72e79_57bf17a3d0fc7.jpg', 'https://storage6.openstreetcam.org/files/photo/2017/11/10/th/966533_dfcfa_13.jpg', 'https://storage6.openstreetcam.org/files/photo/2017/11/10/th/966541_6cb2b_25.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/7/26/th/2818418_83ecb_5f1d8e994f5a5.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/7/28/th/2827634_9008c_5f20513eb1ff1.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/8/5/th/2865646_84ea2_5f2a2bff88b76.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/8/5/th/2865742_9bb8d_5f2a3576d6d95.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/8/15/th/2917218_41295_5f37ec1e71d2e.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/8/23/th/2943810_b2100_5f41d7658e0a9.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/10/28/th/3196014_31e7b_5f998cea88f8f.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/10/29/th/3198870_89d90_5f9b0643d292d.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/10/29/th/3199162_7a9e2_5f9b081742a14.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/1/th/3205238_64ceb_5f9e798e2979d.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/1/th/3205398_14ce2_5f9e7ed544f89.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/1/th/3206602_6b455_5f9eb03161951.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/1/th/3206742_3decc_5f9ec69919c90.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/2/th/3209390_a1167_5fa0601ab7848.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/4/th/3213566_fb0db_5fa27346d9c46.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/4/th/3213734_7853c_5fa2944adf2ef.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/5/th/3215882_0bb67_5fa3cd8e21635.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/5/th/3215958_e71a1_5fa3ce17be096.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/8/th/3224265_8d50e_5fa73e8f77ea8.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/8/th/3224453_ba649_5fa74e036b624.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/12/th/3236665_8e862_5fac90a67954a.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/13/th/3239009_b9362_5fae153ecec5e.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/13/th/3239065_974d3_5fae1924b8065.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/11/29/th/3304101_4cb8b_5fc38c09b7703.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/12/7/th/3316153_b7943_5fcdaece71c88.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/12/7/th/3316185_7407a_5fcdb414af630.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/12/7/th/3317317_308c8_5fce5254dced9.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/12/7/th/3317361_349e8_5fce547e5035e.jpg', 'https://storage13.openstreetcam.org/files/photo/2020/12/7/th/3317421_e7c8e_5fce58f169c2f.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/1/2/th/3344173_7c35a_5ff04b061bcb5.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/1/2/th/3344161_bc71a_5ff04de957f1f.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/1/2/th/3344225_4fc1d_5ff04e9455d52.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/1/2/th/3344221_21e54_5ff054329e35f.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/1/2/th/3345909_ef7e5_5ff0eb74aacb7.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/1/2/th/3345937_9925c_5ff0f06a365c0.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/3/11/th/3455409_cd9b6_604a704e27fc1.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/3/11/th/3455461_ed30a_604a72d5c0776.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/3/12/th/3456997_21395_604be03915e26.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/3/12/th/3457101_98783_604be4c8f093e.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/3/13/th/3459465_131b4_604c70c76ff4e.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/3/13/th/3459813_3bf87_604c98c32365e.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/3/13/th/3459937_6ba6e_604ca69bf0d07.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/3/13/th/3460057_d6ee8_604cad685081d.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/4/10/th/3493841_0b5c6_60716eda00c98.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/4/10/th/3494421_1c859_60719e9b811b7.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/4/10/th/3494453_365eb_6071a1f22f09d.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/4/17/th/3513209_eb480_607b55de4bb18.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/4/17/th/3513249_be607_607b582d19585.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/6/6/th/3640897_74682_60bc324524a85.jpg', 'https://storage13.openstreetcam.org/files/photo/2021/6/6/th/3640957_fdda1_60bc3693dedf0.jpg']\n",
      "https://storage2.openstreetcam.org/files/photo/2016/8/25/th/14682_72e79_57bf17a3d0fc7.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage6.openstreetcam.org/files/photo/2017/11/10/th/966533_dfcfa_13.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage6.openstreetcam.org/files/photo/2017/11/10/th/966541_6cb2b_25.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/7/26/th/2818418_83ecb_5f1d8e994f5a5.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/7/28/th/2827634_9008c_5f20513eb1ff1.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/8/5/th/2865646_84ea2_5f2a2bff88b76.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/8/5/th/2865742_9bb8d_5f2a3576d6d95.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/8/15/th/2917218_41295_5f37ec1e71d2e.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/8/23/th/2943810_b2100_5f41d7658e0a9.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/10/28/th/3196014_31e7b_5f998cea88f8f.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/10/29/th/3198870_89d90_5f9b0643d292d.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/10/29/th/3199162_7a9e2_5f9b081742a14.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/1/th/3205238_64ceb_5f9e798e2979d.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/1/th/3205398_14ce2_5f9e7ed544f89.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/1/th/3206602_6b455_5f9eb03161951.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/1/th/3206742_3decc_5f9ec69919c90.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/2/th/3209390_a1167_5fa0601ab7848.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/4/th/3213566_fb0db_5fa27346d9c46.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/4/th/3213734_7853c_5fa2944adf2ef.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/5/th/3215882_0bb67_5fa3cd8e21635.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/5/th/3215958_e71a1_5fa3ce17be096.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/8/th/3224265_8d50e_5fa73e8f77ea8.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/8/th/3224453_ba649_5fa74e036b624.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/12/th/3236665_8e862_5fac90a67954a.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/13/th/3239009_b9362_5fae153ecec5e.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/13/th/3239065_974d3_5fae1924b8065.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/11/29/th/3304101_4cb8b_5fc38c09b7703.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/12/7/th/3316153_b7943_5fcdaece71c88.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/12/7/th/3316185_7407a_5fcdb414af630.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/12/7/th/3317317_308c8_5fce5254dced9.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/12/7/th/3317361_349e8_5fce547e5035e.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2020/12/7/th/3317421_e7c8e_5fce58f169c2f.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/1/2/th/3344173_7c35a_5ff04b061bcb5.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/1/2/th/3344161_bc71a_5ff04de957f1f.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/1/2/th/3344225_4fc1d_5ff04e9455d52.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/1/2/th/3344221_21e54_5ff054329e35f.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/1/2/th/3345909_ef7e5_5ff0eb74aacb7.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/1/2/th/3345937_9925c_5ff0f06a365c0.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/3/11/th/3455409_cd9b6_604a704e27fc1.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/3/11/th/3455461_ed30a_604a72d5c0776.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/3/12/th/3456997_21395_604be03915e26.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/3/12/th/3457101_98783_604be4c8f093e.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/3/13/th/3459465_131b4_604c70c76ff4e.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/3/13/th/3459813_3bf87_604c98c32365e.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/3/13/th/3459937_6ba6e_604ca69bf0d07.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/3/13/th/3460057_d6ee8_604cad685081d.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/4/10/th/3493841_0b5c6_60716eda00c98.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/4/10/th/3494421_1c859_60719e9b811b7.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/4/10/th/3494453_365eb_6071a1f22f09d.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/4/17/th/3513209_eb480_607b55de4bb18.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/4/17/th/3513249_be607_607b582d19585.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/6/6/th/3640897_74682_60bc324524a85.jpg\n",
      "Image downloaded successfully.\n",
      "https://storage13.openstreetcam.org/files/photo/2021/6/6/th/3640957_fdda1_60bc3693dedf0.jpg\n",
      "Image downloaded successfully.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Define the API endpoint with the specific latitude and longitude\n",
    "# 13.38953209\n",
    "base_url = \"https://api.openstreetcam.org/2.0/photo/?lat={}&lng={}\"\n",
    "\n",
    "\n",
    "# Send a GET request to the API\n",
    "response = requests.get(url)\n",
    "\n",
    "# Check if the request was successful\n",
    "if response.status_code == 200:\n",
    "    # Debug: Print the response JSON to find the correct key for the image URL\n",
    "    #print(response.json())\n",
    "\n",
    "    data = response.json()\n",
    "\n",
    "    # Define a regular expression pattern to match the URLs\n",
    "    url_pattern = r'https://storage\\d+\\.openstreetcam\\.org/files/photo/\\d+/\\d+/\\d+/[^\"]+\\.jpg'\n",
    "\n",
    "    # Find all matches of the pattern\n",
    "    urls = re.findall(url_pattern, json.dumps(data))\n",
    "    #print(urls)\n",
    "\n",
    "    # Filter URLs that include \"{{sizeprefix}}\", \"proc\", or \"th\" before the last \"/\" sign\n",
    "    filtered_urls = []\n",
    "    for url in urls:\n",
    "    # Split the URL by \"/\" to get all segments\n",
    "        parts = url.split('/')\n",
    "    # Check if the segment before the last one matches any of the specified patterns\n",
    "        if len(parts) > 1:  # Ensure there are enough parts to inspect\n",
    "            segment_to_check = parts[-2]  # Get the segment before the last one\n",
    "        # Check if the segment does not contain any of the specified patterns\n",
    "            if not any(pattern in segment_to_check for pattern in [\"{{sizeprefix}}\", \"proc\"]) and (segment_to_check.endswith(\"th\") and not segment_to_check.endswith(\"lth\")):\n",
    "                filtered_urls.append(url)\n",
    "    \n",
    "    print(filtered_urls)\n",
    "    \n",
    "    if filtered_urls:  \n",
    "        # Check if image_url is not None\n",
    "        for image_url in filtered_urls:\n",
    "            print(image_url)\n",
    "            file_name = image_url.split('/')[-1]\n",
    "            if not file_name.endswith('.jpg'):\n",
    "                file_name += '.jpg'\n",
    "        # Combine the directory with the filename\n",
    "            file_path = os.path.join(current_directory, file_name)\n",
    "        # Send a GET request to fetch the image\n",
    "            image_response = requests.get(image_url)\n",
    "    \n",
    "            if image_response.status_code == 200:\n",
    "            # Save the image to a file\n",
    "                with open(file_path, 'wb') as f:\n",
    "                    f.write(image_response.content)\n",
    "                print(\"Image downloaded successfully.\")\n",
    "            else:\n",
    "                print(\"Failed to download the image.\")\n",
    "    else:\n",
    "        print(\"Image URL was not found in the API response.\")\n",
    "else:\n",
    "    print(\"Failed to retrieve data from the API.\")\n",
    "\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "ufo-predict2 (Tensorflow 2.11.0)",
   "language": "python",
   "name": "kai-ufo-predict2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
